WIP: Update rust syntax and integrate travis

suniltheta opened the pull request:
Updated rust syntax. Add many new constructors. Modified code to use new
constructor. Integrated travis CI for the repository.

suniltheta commented on src/lib.rs:
```diff
@@ -1863,87 +1723,89 @@ mod test {
     }
 
     #[test]
+    #[ignore]
```
Currently, I ignored only this test case for pull request sake. We are not yet successful with respect to getting deserializer work for this particular enum CacheState.
Can you please guide us how to move forward with this.

jdm replied:
What output do you get right now?

suniltheta replied:
---- test::test_cache_unknown_entries stdout ----
        thread 'test::test_cache_unknown_entries' panicked at 'called `Result::unwrap()` on an `Err` value: Syntax(Message("missing field `beforeRequest`"), 1, 2)', C:\bot\slave\stable-dist-rustc-win-msvc-64\build\src\libcore\result.rs:868

suniltheta replied:
It expects JSON to be of format 
`Cache { 
before_request: Unknown, 
after_request: Unknown, 
comment: None 
}`
Not empty object.

jdm replied:
I think you can use the `default` annotation for the fields of `Cache` as well to avoid this.

kbrey replied:
Fixed- thanks for the tip about default!

jdm commented on .travis.yml:
```diff
@@ -0,0 +1,24 @@
+
+env:
+  - BUILD_KIND=DEBUG
+  - BUILD_KIND=RELEASE
```
I don't think this doesn't anything? I don't think there's any need to do separate debug and release builds for this crate, in any case.

suniltheta replied:
done

jdm commented on .travis.yml:
```diff
@@ -0,0 +1,24 @@
+rust:
+  - stable
+  - beta
+  - nightly
```
Let's remove nightly from this list; it seems like overkill.

jdm commented on .travis.yml:
```diff
@@ -0,0 +1,24 @@
+  - nightly
+matrix:
+  allow_failures:
+    - rust: nightly
```
We won't need this if we remove nightly builds.

jdm commented on .travis.yml:
```diff
@@ -0,0 +1,24 @@
+
+os:
+  - linux
+  - osx
```
I think it's overkill to do OS X builds when none of the code in this crate is platform-dependent.

suniltheta commented:
I have added .travis.yml file which will trigger build on every pull request merged. Please review the .travis.yml file and let us know if all the build components are required.
Also, let us know if Travis integration is enabled for servo/rust-har

jdm commented on src/lib.rs:
```diff
@@ -758,45 +626,67 @@ pub struct Cache {
 }
 
+/*
+use serde::de::{Deserialize, Deserializer, Visitor, Error};
```
Let's remove this commented out code.

suniltheta replied:
done

jdm commented on src/lib.rs:
```diff
@@ -808,58 +698,41 @@ pub struct CacheEntry {
 
 /// A timing value which may be absent or present
 ///
 /// Defaults to -1 in the absent case.
```
Have we verified that this is still true? Is there a test for it?

suniltheta replied:
I think I have made a mistake here :( . The spec says that "If a time info is not available appropriate field is set to -1". But in my test cases, I am verifying whether those OptionalTiming.NotApplicable (enum)  are 'null' and then passing them.  ( I changed all -1 to null so that the test cases pass).
Should I write a deserializer to make that default value is '-1' not 'null' ??

jdm replied:
Yeah, I think a custom deserializer will be necessary.

jdm replied:
Actually, Serde may support a `default` annotation that could avoid the need for a custom deserializer: http://erickt.github.io/blog/2016/02/26/serde-0-dot-7/

kbrey replied:
I think that default only handles the case where the value is missing -- " returns a default value for a field if it’s not present when deserializing". In this case since the value is there (-1) and we want to map it to a different enum value (NotApplicable), I'm thinking we will have to write a custom deserializer. Please correct me if there's a better option!

jdm replied:
I think you are right.

jdm commented on src/lib.rs:
```diff
@@ -1065,29 +945,33 @@ mod test {
                         }";
-        assert_eq!(json::from_str(log_json),
-                   json::from_str(json::encode(&log).as_slice()));
+		let log_from_str: Log = serde_json::from_str(log_json).unwrap();
```
nit: the indentation here is weird; I would expect it to match the assert_eq on the next line.

suniltheta replied:
done

jdm commented on src/lib.rs:
```diff
@@ -1065,29 +945,33 @@ mod test {
-        assert_eq!(json::from_str(log_json),
-                   json::from_str(json::encode(&log).as_slice()));
+		let log_from_str: Log = serde_json::from_str(log_json).unwrap();
+        assert_eq!( log_from_str, log );
```
nit: remove the space after ( and before ). Same for other instances in this file.

suniltheta replied:
done

jdm commented on src/lib.rs:
```diff
@@ -368,49 +271,50 @@ pub struct Request {
 
 /// This object contains detailed info about the response.
+#[derive(Serialize, Deserialize, PartialEq, Debug)]
+#[serde(rename_all = "camelCase")]
```
The [specification](http://www.softwareishard.com/blog/har-12-spec/#response) says that the redirect_url field should be redirectURL in the final JSON, but this turns it into redirectUrl.

suniltheta replied:
Yes, this was turning redirect_url into redirectUrl. And that's what I verified in 5 test cases including "test_response()".  But we have to write a deserializer which can turn redirect_url to redirectURL (not camelCase redirectUrl).

kbrey replied:
I don't think we need a custom deserializer in this case- there's an option for attribute specific naming which I will use here.

kbrey replied:
Fixed.

jdm commented on src/lib.rs:
```diff
@@ -1863,87 +1723,89 @@ mod test {
+            "123456789".to_string(), 
+            42, 
+            None
+        );         
```
nit: remove the trailing spaces on this and other lines.

suniltheta replied:
done

jdm reviewed (changes requested):
Great work!

jdm commented on src/lib.rs:
```diff
@@ -693,14 +703,30 @@ impl CacheEntry {
+    {
+        let deser_result: serde_json::Value = try!(serde::Deserialize::deserialize(deserializer));
+        match deser_result {
+            //serde_json::Value::Number(n) => Ok(OptionalTiming::TimedContent(n.as_u64().unwrap() as u32)),
```
Let's remove this commented out code.

suniltheta replied:
done

jdm commented on src/lib.rs:
```diff
@@ -693,14 +703,30 @@ impl CacheEntry {
+        let deser_result: serde_json::Value = try!(serde::Deserialize::deserialize(deserializer));
+        match deser_result {
+            //serde_json::Value::Number(n) => Ok(OptionalTiming::TimedContent(n.as_u64().unwrap() as u32)),
+            serde_json::Value::Number(ref n) if n.as_i64().unwrap() >= 0 as i64 => 
```
Will this panic if there is non-i64 content like `5.3` in the JSON for this field? Also, the `as i64` cast shouldn't be necessary here.

suniltheta replied:
Added a test case so that test_page_timings panics on float value. And made a slight modification on Deserialize for OptionalTiming (Thanks for inputs from Katie).

jdm commented on src/lib.rs:
```diff
@@ -693,14 +703,30 @@ impl CacheEntry {
+            //serde_json::Value::Number(n) => Ok(OptionalTiming::TimedContent(n.as_u64().unwrap() as u32)),
+            serde_json::Value::Number(ref n) if n.as_i64().unwrap() >= 0 as i64 => 
+	            Ok(OptionalTiming::TimedContent(n.as_u64().unwrap() as u32)),
+            serde_json::Value::Number(ref n) if n.as_i64().unwrap() == -1 as i64 => 
```
The `as i64` cast shouldn't be necessary here.

suniltheta replied:
done

jdm reviewed (changes requested):
Good improvements!

jdm commented on src/lib.rs:
```diff
@@ -1098,6 +1098,21 @@ mod test {
     }
 
     #[test]
+    #[should_panic(expected = "Unexpected value")]
```
My earlier comment was pointing out that panicking (due to `unwrap()`) is not the behaviour we want, so this test should not be expecting it.

jdm replied:
Rather than calling `unwrap` on the result of `from_str`, let's assert that the result `is_err()` instead.

suniltheta replied:
We are not expecting a panic due to unwrap() here. We are expecting a panic for not giving integer value. Should this change?

jdm replied:
The panic is occurring because this test is calling `unwrap` on an Err value, which is what is returned by the deserializer when an integer is not present.

suniltheta replied:
It has been fixed now.

suniltheta replied:
Sorry for confusion, 
After my last commit I am asserting 'page_timings_result.is_err()'
You can see this change in 'test_page_timings_float' Test

sunilthetab committed:
Update rust syntax and integrate travis

Updated rust syntax. Add many new constructors. Modified code to use new
constructor. Integrated travis CI for the repository.

suniltheta commented:
@jdm I have fixed all issues and squashed commits. Please let me know if any changes are required. Thanks!

jdm commented:
Thank you!

suniltheta commented:
Just one other change to be done from servo side,
The Travis badge on README.md should point to https://travis-ci.org/servo/rust-har instead of https://travis-ci.org/suniltheta/rust-har
